LocalMultiplayerAgent/DeploymentTool/DeploymentScript.cs

```diff
@@ -0,0 +1,270 @@
+                OsPlatform = deploymentSettings.OSPlatform
+            };
+        }
+
+        public void PrintDeploymentMessage(string buildName, List<BuildRegionParams> regionConfigurations)
+        {
+            Console.WriteLine($"Starting deployment {buildName} for titleId, regions  {string.Join(", ", regionConfigurations.Select(x => x.Region))}");
+        }
+
+        public async Task<PlayFabResult<CreateBuildWithProcessBasedServerResponse>> CreateBuildWithProcessBasedServer(CreateBuildWithProcessBasedServerRequest request)
+        {
+            PrintDeploymentMessage(request.BuildName, request.RegionConfigurations);
+
+            return await PlayFabMultiplayerAPI.CreateBuildWithProcessBasedServerAsync(request);
+        }
+
+        public async Task<PlayFabResult<CreateBuildWithManagedContainerResponse>> CreateBuildWithManagedContainer(CreateBuildWithManagedContainerRequest request)
+        {
+            PrintDeploymentMessage(request.BuildName, request.RegionConfigurations);
+
+            return await PlayFabMultiplayerAPI.CreateBuildWithManagedContainerAsync(request);
+        }
+
+        public async Task<PlayFabResult<CreateBuildWithCustomContainerResponse>> CreateBuildWithCustomContainer(CreateBuildWithCustomContainerRequest request)
+        {
+            PrintDeploymentMessage(request.BuildName, request.RegionConfigurations);
+
+            return await PlayFabMultiplayerAPI.CreateBuildWithCustomContainerAsync(request);
+        }
+
+        public List<PlayFab.MultiplayerModels.Port> PortMapping()
+        {
+            var ports = new List<Port>();
+
+            foreach (var portList in settings.PortMappingsList)
+            {
+                ports.AddRange(portList?.Select(x => new Port()
+                {
+                    Name = x.GamePort.Name,
+                    Num = settings.RunContainer ? x.GamePort.Number : 0,
+                    Protocol = (ProtocolType)Enum.Parse(typeof(ProtocolType), x.GamePort.Protocol)
+                }).ToList());
+            }
+
+            return ports;
+        }
+
+        public string GetAssetFileNameFromPath(string filePath)
+        {
+            return Path.GetFileName(filePath);
+        }
+
+        public async Task<PlayFabResult<GetAssetDownloadUrlResponse>> FileExistsInBlob(string filename)
+        {
+            GetAssetDownloadUrlRequest downloadRequest = new GetAssetDownloadUrlRequest() { FileName = filename };
+
+            return await PlayFabMultiplayerAPI.GetAssetDownloadUrlAsync(downloadRequest);
+        }
+
+        public void PrintError(dynamic error)
+        {
+            if (error != null && error.ErrorMessage != null)
+            {
+                Console.WriteLine($"{error.ErrorMessage}");
+            }
+        }
+
+        public async Task CheckAssetFiles(string filename)
+        {
+           
+            var filevalidator = FileExistsInBlob(filename);
```

Comment:
If I'm understanding this correctly, you're sending a request to ControlPlane to see if an asset already exists before attempting to upload it--could you instead just attempt to upload it and ControlPlane will return an error if the asset already exists?

Reply:
Also, does this allow developer to reuse existing assets? Like, is there a way for them to create a build that uses assets they've already uploaded to a previous build?

Reply:
Yes, we can attempt to upload it and ControlPlane will return an error if the asset already exists. I've update. I'm also curious what your thinking is behind this. I'd think less async API calls is always better, i.e. more efficient, cheaper

Reply:
Fewer API calls is technically cheaper, but this is an extremely low traffic API--I'd be surprised if we got more than 100 asset uploads per day. And duplicating the validation logic that guarantees a duplicate Asset can't be uploaded just makes it harder to maintain down the road--if for some reason we changed how we validated that (maybe in the future we allow duplicate asset names?) we'd have 2 places that need changing.